File: src/rate_limiter.zig
//! Rate Limiting and DDoS Protection for Wraith
//! Implements token bucket and sliding window algorithms with injected crypto for secure hashing

const std = @import("std");
const root = @import("root.zig");
const crypto = root.crypto_interface;

const Allocator = std.mem.Allocator;

pub const RateLimiter = struct {
    allocator: Allocator,
    config: RateLimitConfig,
    client_buckets: std.StringHashMap(TokenBucket),
    global_bucket: TokenBucket,
    blocked_ips: std.StringHashMap(BlockedClient),
    
    const Self = @This();
    
    pub fn init(allocator: Allocator, config: RateLimitConfig) Self {
        return Self{
            .allocator = allocator,
            .config = config,
            .client_buckets = std.StringHashMap(TokenBucket).init(allocator),
            .global_bucket = TokenBucket.init(config.global_requests_per_second * 60, config.global_burst),
            .blocked_ips = std.StringHashMap(BlockedClient).init(allocator),
        };
    }
    
    pub fn deinit(self: *Self) void {
        var bucket_iter = self.client_buckets.iterator();
        while (bucket_iter.next()) |entry| {
            self.allocator.free(entry.key_ptr.*);
        }
        self.client_buckets.deinit();
        
        var blocked_iter = self.blocked_ips.iterator();
        while (blocked_iter.next()) |entry| {
            self.allocator.free(entry.key_ptr.*);
        }
        self.blocked_ips.deinit();
    }
    
    pub fn isAllowed(self: *Self, client_ip: []const u8, request_size: u32) !RateLimitResult {
        const now = std.time.milliTimestamp();
        
        // Check if IP is explicitly blocked
        if (self.blocked_ips.get(client_ip)) |blocked| {
            if (now < blocked.unblock_time) {
                return RateLimitResult{
                    .allowed = false,
                    .reason = .blocked,
                    .retry_after = @intCast((blocked.unblock_time - now) / 1000),
                    .remaining_requests = 0,
                };
            } else {
                // Remove expired block
                _ = self.blocked_ips.remove(client_ip);
            }
        }
        
        // Check whitelist
        for (self.config.whitelist) |whitelisted_ip| {
            if (std.mem.eql(u8, client_ip, whitelisted_ip)) {
                return RateLimitResult{
                    .allowed = true,
                    .reason = .whitelisted,
                    .retry_after = 0,
                    .remaining_requests = std.math.maxInt(u32),
                };
            }
        }
        
        // Check blacklist
        for (self.config.blacklist) |blacklisted_ip| {
            if (std.mem.eql(u8, client_ip, blacklisted_ip)) {
                return RateLimitResult{
                    .allowed = false,
                    .reason = .blacklisted,
                    .retry_after = std.math.maxInt(u32),
                    .remaining_requests = 0,
                };
            }
        }
        
        // Global rate limiting
        if (!self.global_bucket.tryConsume(1, now)) {
            return RateLimitResult{
                .allowed = false,
                .reason = .global_limit,
                .retry_after = @intCast(self.global_bucket.getRefillTime(now) / 1000),
                .remaining_requests = 0,
            };
        }
        
        // Per-client rate limiting
        const client_key = try self.allocator.dupe(u8, client_ip);
        var bucket = self.client_buckets.get(client_ip) orelse blk: {
            const new_bucket = TokenBucket.init(
                self.config.requests_per_minute,
                self.config.burst
            );
            try self.client_buckets.put(client_key, new_bucket);
            break :blk new_bucket;
        };
        
        if (!bucket.tryConsume(1, now)) {
            // Check if this client should be temporarily blocked
            if (self.shouldBlockClient(client_ip, now)) {
                try self.blockClient(client_ip, now);
                return RateLimitResult{
                    .allowed = false,
                    .reason = .blocked,
                    .retry_after = self.config.block_duration,
                    .remaining_requests = 0,
                };
            }
            
            return RateLimitResult{
                .allowed = false,
                .reason = .rate_limited,
                .retry_after = @intCast(bucket.getRefillTime(now) / 1000),
                .remaining_requests = bucket.tokens,
            };
        }
        
        // Check request size limits
        if (request_size > self.config.max_request_size) {
            return RateLimitResult{
                .allowed = false,
                .reason = .request_too_large,
                .retry_after = 0,
                .remaining_requests = bucket.tokens,
            };
        }
        
        // Update bucket in map
        try self.client_buckets.put(client_key, bucket);
        
        return RateLimitResult{
            .allowed = true,
            .reason = .allowed,
            .retry_after = 0,
            .remaining_requests = bucket.tokens,
        };
    }
    
    fn shouldBlockClient(self: *Self, client_ip: []const u8, now: i64) bool {
        _ = client_ip;
        _ = now;
        
        // Simple heuristic: block if client has been rate limited multiple times
        // In a real implementation, this would track violation history
        return self.config.auto_block_enabled;
    }
    
    fn blockClient(self: *Self, client_ip: []const u8, now: i64) !void {
        const blocked_client = BlockedClient{
            .ip = try self.allocator.dupe(u8, client_ip),
            .block_time = now,
            .unblock_time = now + (self.config.block_duration * 1000),
            .reason = "Rate limit violations",
        };
        
        try self.blocked_ips.put(blocked_client.ip, blocked_client);
        std.log.warn("Blocked client {} for {} seconds", .{ client_ip, self.config.block_duration });
    }
    
    pub fn cleanupExpiredEntries(self: *Self) void {
        const now = std.time.milliTimestamp();
        
        // Clean up expired blocked IPs
        var blocked_iter = self.blocked_ips.iterator();
        var to_remove = std.ArrayList([]const u8).init(self.allocator);
        defer to_remove.deinit();
        
        while (blocked_iter.next()) |entry| {
            if (now >= entry.value_ptr.unblock_time) {
                to_remove.append(entry.key_ptr.*) catch continue;
            }
        }
        
        for (to_remove.items) |ip| {
            if (self.blocked_ips.fetchRemove(ip)) |removed| {
                self.allocator.free(removed.key);
                self.allocator.free(removed.value.ip);
            }
        }
        
        // Clean up old client buckets (optional - could implement LRU)
        // For now, keep all buckets to maintain rate limiting state
    }
    
    pub fn addToBlacklist(self: *Self, ip: []const u8) !void {
        // In a real implementation, this would persist to configuration
        std.log.info("Added {} to blacklist", .{ip});
        _ = self;
    }
    
    pub fn removeFromBlacklist(self: *Self, ip: []const u8) !void {
        // In a real implementation, this would persist to configuration
        std.log.info("Removed {} from blacklist", .{ip});
        _ = self;
    }
    
    pub fn getStats(self: *Self) RateLimitStats {
        var blocked_count: u32 = 0;
        var active_clients: u32 = 0;
        
        const now = std.time.milliTimestamp();
        
        var blocked_iter = self.blocked_ips.iterator();
        while (blocked_iter.next()) |entry| {
            if (now < entry.value_ptr.unblock_time) {
                blocked_count += 1;
            }
        }
        
        active_clients = @intCast(self.client_buckets.count());
        
        return RateLimitStats{
            .active_clients = active_clients,
            .blocked_clients = blocked_count,
            .global_tokens_remaining = self.global_bucket.tokens,
            .total_blocks = blocked_count, // Simplified
        };
    }
};

pub const TokenBucket = struct {
    capacity: u32,
    tokens: u32,
    refill_rate: u32, // tokens per minute
    last_refill: i64, // milliseconds
    
    const Self = @This();
    
    pub fn init(capacity: u32, refill_rate: u32) Self {
        return Self{
            .capacity = capacity,
            .tokens = capacity,
            .refill_rate = refill_rate,
            .last_refill = std.time.milliTimestamp(),
        };
    }
    
    pub fn tryConsume(self: *Self, tokens_needed: u32, now: i64) bool {
        self.refill(now);
        
        if (self.tokens >= tokens_needed) {
            self.tokens -= tokens_needed;
            return true;
        }
        
        return false;
    }
    
    fn refill(self: *Self, now: i64) void {
        const time_passed = now - self.last_refill;
        if (time_passed <= 0) return;
        
        // Calculate tokens to add (rate is per minute)
        const tokens_to_add = @as(u32, @intCast((time_passed * self.refill_rate) / 60000));
        
        if (tokens_to_add > 0) {
            self.tokens = @min(self.capacity, self.tokens + tokens_to_add);
            self.last_refill = now;
        }
    }
    
    pub fn getRefillTime(self: *Self, now: i64) i64 {
        if (self.tokens >= self.capacity) return 0;
        
        const tokens_needed = self.capacity - self.tokens;
        const time_needed = (tokens_needed * 60000) / self.refill_rate;
        
        return time_needed;
    }
};

pub const DDoSProtector = struct {
    allocator: Allocator,
    config: DDoSConfig,
    connection_tracker: std.StringHashMap(ConnectionInfo),
    packet_tracker: std.StringHashMap(PacketInfo),
    
    const Self = @This();
    
    pub fn init(allocator: Allocator, config: DDoSConfig) Self {
        return Self{
            .allocator = allocator,
            .config = config,
            .connection_tracker = std.StringHashMap(ConnectionInfo).init(allocator),
            .packet_tracker = std.StringHashMap(PacketInfo).init(allocator),
        };
    }
    
    pub fn deinit(self: *Self) void {
        var conn_iter = self.connection_tracker.iterator();
        while (conn_iter.next()) |entry| {
            self.allocator.free(entry.key_ptr.*);
        }
        self.connection_tracker.deinit();
        
        var packet_iter = self.packet_tracker.iterator();
        while (packet_iter.next()) |entry| {
            self.allocator.free(entry.key_ptr.*);
        }
        self.packet_tracker.deinit();
    }
    
    pub fn checkConnection(self: *Self, client_ip: []const u8) !bool {
        if (!self.config.enabled) return true;
        
        const now = std.time.milliTimestamp();
        const client_key = try self.allocator.dupe(u8, client_ip);
        
        var info = self.connection_tracker.get(client_ip) orelse ConnectionInfo{
            .connections = 0,
            .last_connection = now,
            .first_connection = now,
        };
        
        // Reset counter if window has passed
        if (now - info.first_connection > self.config.window_size * 1000) {
            info.connections = 0;
            info.first_connection = now;
        }
        
        info.connections += 1;
        info.last_connection = now;
        
        const allowed = info.connections <= self.config.max_connections_per_ip;
        
        try self.connection_tracker.put(client_key, info);
        
        if (!allowed) {
            std.log.warn("DDoS protection: Too many connections from {s}", .{client_ip});
        }
        
        return allowed;
    }
    
    pub fn checkPacketRate(self: *Self, client_ip: []const u8) !bool {
        if (!self.config.enabled) return true;
        
        const now = std.time.milliTimestamp();
        const client_key = try self.allocator.dupe(u8, client_ip);
        
        var info = self.packet_tracker.get(client_ip) orelse PacketInfo{
            .packets = 0,
            .last_packet = now,
            .first_packet = now,
        };
        
        // Reset counter if window has passed
        if (now - info.first_packet > 1000) { // 1 second window
            info.packets = 0;
            info.first_packet = now;
        }
        
        info.packets += 1;
        info.last_packet = now;
        
        const allowed = info.packets <= self.config.packet_rate_limit;
        
        try self.packet_tracker.put(client_key, info);
        
        if (!allowed) {
            std.log.warn("DDoS protection: Packet rate limit exceeded for {s}", .{client_ip});
        }
        
        return allowed;
    }
    
    pub fn cleanupExpiredEntries(self: *Self) void {
        const now = std.time.milliTimestamp();
        const window_ms = self.config.window_size * 1000;
        
        // Clean up connection tracker
        var conn_iter = self.connection_tracker.iterator();
        var conn_to_remove = std.ArrayList([]const u8).init(self.allocator);
        defer conn_to_remove.deinit();
        
        while (conn_iter.next()) |entry| {
            if (now - entry.value_ptr.last_connection > window_ms) {
                conn_to_remove.append(entry.key_ptr.*) catch continue;
            }
        }
        
        for (conn_to_remove.items) |ip| {
            if (self.connection_tracker.fetchRemove(ip)) |removed| {
                self.allocator.free(removed.key);
            }
        }
        
        // Clean up packet tracker
        var packet_iter = self.packet_tracker.iterator();
        var packet_to_remove = std.ArrayList([]const u8).init(self.allocator);
        defer packet_to_remove.deinit();
        
        while (packet_iter.next()) |entry| {
            if (now - entry.value_ptr.last_packet > 10000) { // 10 second cleanup
                packet_to_remove.append(entry.key_ptr.*) catch continue;
            }
        }
        
        for (packet_to_remove.items) |ip| {
            if (self.packet_tracker.fetchRemove(ip)) |removed| {
                self.allocator.free(removed.key);
            }
        }
    }
};

// Configuration structures
pub const RateLimitConfig = struct {
    enabled: bool = true,
    requests_per_minute: u32 = 60,
    burst: u32 = 10,
    global_requests_per_second: u32 = 1000,
    global_burst: u32 = 100,
    max_request_size: u32 = 1024 * 1024, // 1MB
    auto_block_enabled: bool = true,
    block_duration: u32 = 300, // 5 minutes
    whitelist: []const []const u8 = &.{},
    blacklist: []const []const u8 = &.{},
};

pub const DDoSConfig = struct {
    enabled: bool = true,
    max_connections_per_ip: u32 = 100,
    connection_rate_limit: u32 = 10, // per second
    packet_rate_limit: u32 = 1000, // per second
    window_size: u32 = 60, // seconds
};

// Result and info structures
pub const RateLimitResult = struct {
    allowed: bool,
    reason: RateLimitReason,
    retry_after: u32, // seconds
    remaining_requests: u32,
};

pub const RateLimitReason = enum {
    allowed,
    whitelisted,
    rate_limited,
    blacklisted,
    blocked,
    global_limit,
    request_too_large,
};

pub const BlockedClient = struct {
    ip: []const u8,
    block_time: i64,
    unblock_time: i64,
    reason: []const u8,
};

pub const ConnectionInfo = struct {
    connections: u32,
    last_connection: i64,
    first_connection: i64,
};

pub const PacketInfo = struct {
    packets: u32,
    last_packet: i64,
    first_packet: i64,
};

pub const RateLimitStats = struct {
    active_clients: u32,
    blocked_clients: u32,
    global_tokens_remaining: u32,
    total_blocks: u32,
};

/// Create a combined security manager with rate limiting and DDoS protection
pub const SecurityManager = struct {
    allocator: Allocator,
    rate_limiter: RateLimiter,
    ddos_protector: DDoSProtector,
    
    const Self = @This();
    
    pub fn init(allocator: Allocator, rate_config: RateLimitConfig, ddos_config: DDoSConfig) Self {
        return Self{
            .allocator = allocator,
            .rate_limiter = RateLimiter.init(allocator, rate_config),
            .ddos_protector = DDoSProtector.init(allocator, ddos_config),
        };
    }
    
    pub fn deinit(self: *Self) void {
        self.rate_limiter.deinit();
        self.ddos_protector.deinit();
    }
    
    pub fn checkRequest(self: *Self, client_ip: []const u8, request_size: u32) !SecurityResult {
        // Check DDoS protection first
        if (!(try self.ddos_protector.checkConnection(client_ip))) {
            return SecurityResult{
                .allowed = false,
                .reason = "DDoS protection: too many connections",
                .rate_limit_result = null,
            };
        }
        
        if (!(try self.ddos_protector.checkPacketRate(client_ip))) {
            return SecurityResult{
                .allowed = false,
                .reason = "DDoS protection: packet rate limit exceeded",
                .rate_limit_result = null,
            };
        }
        
        // Check rate limiting
        const rate_result = try self.rate_limiter.isAllowed(client_ip, request_size);
        
        return SecurityResult{
            .allowed = rate_result.allowed,
            .reason = if (rate_result.allowed) "allowed" else @tagName(rate_result.reason),
            .rate_limit_result = rate_result,
        };
    }
    
    pub fn cleanup(self: *Self) void {
        self.rate_limiter.cleanupExpiredEntries();
        self.ddos_protector.cleanupExpiredEntries();
    }
};

pub const SecurityResult = struct {
    allowed: bool,
    reason: []const u8,
    rate_limit_result: ?RateLimitResult,
};

File: src/rate_limiter.rs
use crate::config::{RateLimitConfig, DdosConfig};
use anyhow::Result;
use governor::{Quota, RateLimiter as GovernorRateLimiter, DefaultDirectRateLimiter};
use nonzero_ext::*;
use std::collections::HashMap;
use std::net::IpAddr;
use std::sync::Arc;
use std::time::{Duration, Instant};
use tokio::sync::RwLock;
use tracing::{debug, warn, info};

#[derive(Clone)]
pub struct RateLimiter {
    config: RateLimitConfig,
    ddos_config: DdosConfig,
    global_limiter: Arc<DefaultDirectRateLimiter>,
    per_ip_limiters: Arc<RwLock<HashMap<IpAddr, Arc<DefaultDirectRateLimiter>>>>,
    blocked_ips: Arc<RwLock<HashMap<IpAddr, BlockedClient>>>,
    connection_counts: Arc<RwLock<HashMap<IpAddr, ConnectionTracker>>>,
}

#[derive(Debug, Clone)]
struct BlockedClient {
    blocked_until: Instant,
    reason: BlockReason,
    block_count: u32,
}

#[derive(Debug, Clone)]
enum BlockReason {
    RateLimit,
    TooManyConnections,
    DdosDetection,
    Blacklisted,
}

#[derive(Debug, Clone)]
struct ConnectionTracker {
    active_connections: u32,
    last_connection: Instant,
    connection_rate: Vec<Instant>,
}

#[derive(Debug)]
pub struct RateLimitResult {
    pub allowed: bool,
    pub reason: RateLimitReason,
    pub retry_after: Option<Duration>,
    pub remaining: Option<u32>,
}

#[derive(Debug, Clone)]
pub enum RateLimitReason {
    Allowed,
    RateLimit,
    GlobalLimit,
    Blocked,
    TooManyConnections,
    Blacklisted,
    Whitelisted,
}

impl RateLimiter {
    pub fn new(rate_config: RateLimitConfig, ddos_config: DdosConfig) -> Self {
        // Create global rate limiter
        let global_quota = Quota::per_minute(nonzero!(rate_config.requests_per_minute))
            .allow_burst(nonzero!(rate_config.burst));
        let global_limiter = Arc::new(DefaultDirectRateLimiter::new(global_quota));

        Self {
            config: rate_config,
            ddos_config,
            global_limiter,
            per_ip_limiters: Arc::new(RwLock::new(HashMap::new())),
            blocked_ips: Arc::new(RwLock::new(HashMap::new())),
            connection_counts: Arc::new(RwLock::new(HashMap::new())),
        }
    }

    pub async fn check_request(
        &self,
        client_ip: IpAddr,
        request_size: Option<usize>,
    ) -> Result<RateLimitResult> {
        if !self.config.enabled {
            return Ok(RateLimitResult {
                allowed: true,
                reason: RateLimitReason::Allowed,
                retry_after: None,
                remaining: None,
            });
        }

        // Check if IP is whitelisted
        if self.is_whitelisted(client_ip) {
            return Ok(RateLimitResult {
                allowed: true,
                reason: RateLimitReason::Whitelisted,
                retry_after: None,
                remaining: None,
            });
        }

        // Check if IP is blacklisted
        if self.is_blacklisted(client_ip) {
            return Ok(RateLimitResult {
                allowed: false,
                reason: RateLimitReason::Blacklisted,
                retry_after: None,
                remaining: Some(0),
            });
        }

        // Check if IP is currently blocked
        if let Some(result) = self.check_blocked_ip(client_ip).await {
            return Ok(result);
        }

        // Check request size limits
        if let Some(size) = request_size {
            if size > self.config.max_request_size {
                self.block_ip(client_ip, BlockReason::RateLimit, Duration::from_secs(300)).await;
                return Ok(RateLimitResult {
                    allowed: false,
                    reason: RateLimitReason::RateLimit,
                    retry_after: Some(Duration::from_secs(300)),
                    remaining: Some(0),
                });
            }
        }

        // Check global rate limit
        if let Err(_) = self.global_limiter.check() {
            debug!("Global rate limit exceeded");
            return Ok(RateLimitResult {
                allowed: false,
                reason: RateLimitReason::GlobalLimit,
                retry_after: Some(Duration::from_secs(60)),
                remaining: Some(0),
            });
        }

        // Check per-IP rate limit
        let per_ip_result = self.check_per_ip_limit(client_ip).await?;
        if !per_ip_result.allowed {
            // Auto-block if enabled
            if self.config.auto_block_enabled {
                self.block_ip(client_ip, BlockReason::RateLimit, self.config.block_duration).await;
            }
            return Ok(per_ip_result);
        }

        // Check DDoS protection
        if self.ddos_config.enabled {
            if let Some(result) = self.check_ddos_protection(client_ip).await {
                return Ok(result);
            }
        }

        Ok(RateLimitResult {
            allowed: true,
            reason: RateLimitReason::Allowed,
            retry_after: None,
            remaining: per_ip_result.remaining,
        })
    }

    pub async fn track_connection(&self, client_ip: IpAddr, connected: bool) -> Result<bool> {
        if !self.ddos_config.enabled {
            return Ok(true);
        }

        let mut connection_counts = self.connection_counts.write().await;
        let now = Instant::now();

        let tracker = connection_counts
            .entry(client_ip)
            .or_insert_with(|| ConnectionTracker {
                active_connections: 0,
                last_connection: now,
                connection_rate: Vec::new(),
            });

        if connected {
            tracker.active_connections += 1;
            tracker.last_connection = now;
            tracker.connection_rate.push(now);

            // Clean old connection rate entries
            tracker.connection_rate.retain(|&time| {
                now.duration_since(time) <= Duration::from_secs(60)
            });

            // Check connection limits
            if tracker.active_connections > self.ddos_config.max_connections_per_ip {
                warn!(
                    "IP {} exceeded max connections: {}",
                    client_ip, tracker.active_connections
                );
                self.block_ip(client_ip, BlockReason::TooManyConnections, Duration::from_secs(600)).await;
                return Ok(false);
            }

            // Check connection rate
            if tracker.connection_rate.len() as u32 > self.ddos_config.connection_rate_limit {
                warn!(
                    "IP {} exceeded connection rate: {} connections/min",
                    client_ip,
                    tracker.connection_rate.len()
                );
                self.block_ip(client_ip, BlockReason::DdosDetection, Duration::from_secs(1800)).await;
                return Ok(false);
            }
        } else {
            tracker.active_connections = tracker.active_connections.saturating_sub(1);
        }

        Ok(true)
    }

    async fn check_per_ip_limit(&self, client_ip: IpAddr) -> Result<RateLimitResult> {
        let limiters = self.per_ip_limiters.read().await;

        if let Some(limiter) = limiters.get(&client_ip) {
            match limiter.check() {
                Ok(_) => Ok(RateLimitResult {
                    allowed: true,
                    reason: RateLimitReason::Allowed,
                    retry_after: None,
                    remaining: None, // TODO: Get remaining from governor
                }),
                Err(_) => Ok(RateLimitResult {
                    allowed: false,
                    reason: RateLimitReason::RateLimit,
                    retry_after: Some(Duration::from_secs(60)),
                    remaining: Some(0),
                }),
            }
        } else {
            // Create new limiter for this IP
            drop(limiters);
            let mut limiters = self.per_ip_limiters.write().await;

            let quota = Quota::per_minute(nonzero!(self.config.requests_per_minute))
                .allow_burst(nonzero!(self.config.burst));
            let limiter = Arc::new(DefaultDirectRateLimiter::new(quota));

            // Check the new limiter
            let result = match limiter.check() {
                Ok(_) => RateLimitResult {
                    allowed: true,
                    reason: RateLimitReason::Allowed,
                    retry_after: None,
                    remaining: None,
                },
                Err(_) => RateLimitResult {
                    allowed: false,
                    reason: RateLimitReason::RateLimit,
                    retry_after: Some(Duration::from_secs(60)),
                    remaining: Some(0),
                },
            };

            limiters.insert(client_ip, limiter);
            Ok(result)
        }
    }

    async fn check_blocked_ip(&self, client_ip: IpAddr) -> Option<RateLimitResult> {
        let mut blocked_ips = self.blocked_ips.write().await;

        if let Some(blocked) = blocked_ips.get(&client_ip) {
            if Instant::now() < blocked.blocked_until {
                let retry_after = blocked.blocked_until.duration_since(Instant::now());
                return Some(RateLimitResult {
                    allowed: false,
                    reason: RateLimitReason::Blocked,
                    retry_after: Some(retry_after),
                    remaining: Some(0),
                });
            } else {
                // Block expired, remove it
                blocked_ips.remove(&client_ip);
                info!("Unblocked IP: {}", client_ip);
            }
        }

        None
    }

    async fn check_ddos_protection(&self, client_ip: IpAddr) -> Option<RateLimitResult> {
        let connection_counts = self.connection_counts.read().await;

        if let Some(tracker) = connection_counts.get(&client_ip) {
            if tracker.active_connections > self.ddos_config.max_connections_per_ip {
                return Some(RateLimitResult {
                    allowed: false,
                    reason: RateLimitReason::TooManyConnections,
                    retry_after: Some(Duration::from_secs(60)),
                    remaining: Some(0),
                });
            }
        }

        None
    }

    async fn block_ip(&self, client_ip: IpAddr, reason: BlockReason, duration: Duration) {
        let mut blocked_ips = self.blocked_ips.write().await;
        let blocked_until = Instant::now() + duration;

        let block_count = if let Some(existing) = blocked_ips.get(&client_ip) {
            existing.block_count + 1
        } else {
            1
        };

        blocked_ips.insert(client_ip, BlockedClient {
            blocked_until,
            reason: reason.clone(),
            block_count,
        });

        warn!(
            "Blocked IP {} for {:?} (reason: {:?}, count: {})",
            client_ip,
            duration,
            reason,
            block_count
        );
    }

    fn is_whitelisted(&self, client_ip: IpAddr) -> bool {
        self.config.whitelist.iter().any(|ip_str| {
            ip_str.parse::<IpAddr>().map_or(false, |ip| ip == client_ip)
        })
    }

    fn is_blacklisted(&self, client_ip: IpAddr) -> bool {
        self.config.blacklist.iter().any(|ip_str| {
            ip_str.parse::<IpAddr>().map_or(false, |ip| ip == client_ip)
        })
    }

    pub async fn get_stats(&self) -> HashMap<String, serde_json::Value> {
        let mut stats = HashMap::new();

        let blocked_ips = self.blocked_ips.read().await;
        let connection_counts = self.connection_counts.read().await;
        let per_ip_limiters = self.per_ip_limiters.read().await;

        stats.insert("blocked_ips_count".to_string(), blocked_ips.len().into());
        stats.insert("tracked_ips_count".to_string(), per_ip_limiters.len().into());
        stats.insert("active_connections_count".to_string(),
            connection_counts.values().map(|t| t.active_connections).sum::<u32>().into());

        let blocked_ips_info: Vec<_> = blocked_ips.iter().map(|(ip, blocked)| {
            serde_json::json!({
                "ip": ip.to_string(),
                "blocked_until": blocked.blocked_until.elapsed().as_secs(),
                "reason": format!("{:?}", blocked.reason),
                "block_count": blocked.block_count,
            })
        }).collect();

        stats.insert("blocked_ips".to_string(), blocked_ips_info.into());
        stats
    }

    pub async fn unblock_ip(&self, client_ip: IpAddr) -> bool {
        let mut blocked_ips = self.blocked_ips.write().await;
        if blocked_ips.remove(&client_ip).is_some() {
            info!("Manually unblocked IP: {}", client_ip);
            true
        } else {
            false
        }
    }

    pub async fn cleanup_expired(&self) {
        let now = Instant::now();

        // Clean up expired blocked IPs
        {
            let mut blocked_ips = self.blocked_ips.write().await;
            blocked_ips.retain(|ip, blocked| {
                let should_keep = now < blocked.blocked_until;
                if !should_keep {
                    debug!("Cleaned up expired block for IP: {}", ip);
                }
                should_keep
            });
        }

        // Clean up old connection trackers
        {
            let mut connection_counts = self.connection_counts.write().await;
            connection_counts.retain(|ip, tracker| {
                let should_keep = tracker.active_connections > 0
                    || now.duration_since(tracker.last_connection) < Duration::from_secs(3600);
                if !should_keep {
                    debug!("Cleaned up connection tracker for IP: {}", ip);
                }
                should_keep
            });
        }

        // Clean up old per-IP limiters
        {
            let mut per_ip_limiters = self.per_ip_limiters.write().await;
            let old_count = per_ip_limiters.len();
            // Keep only recent limiters (this is a simple heuristic)
            if per_ip_limiters.len() > 10000 {
                per_ip_limiters.clear();
                debug!("Cleared {} old per-IP limiters", old_count);
            }
        }
    }
}

// Spawn cleanup task
pub fn spawn_cleanup_task(rate_limiter: Arc<RateLimiter>) {
    tokio::spawn(async move {
        let mut interval = tokio::time::interval(Duration::from_secs(300)); // 5 minutes

        loop {
            interval.tick().await;
            rate_limiter.cleanup_expired().await;
        }
    });
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::net::Ipv4Addr;

    #[tokio::test]
    async fn test_rate_limiter_basic() {
        let config = RateLimitConfig {
            enabled: true,
            requests_per_minute: 60,
            burst: 10,
            max_request_size: 1024 * 1024,
            whitelist: vec![],
            blacklist: vec![],
            auto_block_enabled: false,
            block_duration: Duration::from_secs(300),
        };

        let ddos_config = DdosConfig {
            enabled: false,
            max_connections_per_ip: 100,
            connection_rate_limit: 10,
            packet_rate_limit: 1000,
            window_size: Duration::from_secs(60),
        };

        let limiter = RateLimiter::new(config, ddos_config);
        let ip = IpAddr::V4(Ipv4Addr::new(192, 168, 1, 1));

        let result = limiter.check_request(ip, None).await.unwrap();
        assert!(result.allowed);
        assert!(matches!(result.reason, RateLimitReason::Allowed));
    }

    #[tokio::test]
    async fn test_whitelist() {
        let config = RateLimitConfig {
            enabled: true,
            requests_per_minute: 1, // Very low limit
            burst: 1,
            max_request_size: 1024 * 1024,
            whitelist: vec!["192.168.1.1".to_string()],
            blacklist: vec![],
            auto_block_enabled: false,
            block_duration: Duration::from_secs(300),
        };

        let ddos_config = DdosConfig {
            enabled: false,
            max_connections_per_ip: 100,
            connection_rate_limit: 10,
            packet_rate_limit: 1000,
            window_size: Duration::from_secs(60),
        };

        let limiter = RateLimiter::new(config, ddos_config);
        let ip = IpAddr::V4(Ipv4Addr::new(192, 168, 1, 1));

        let result = limiter.check_request(ip, None).await.unwrap();
        assert!(result.allowed);
        assert!(matches!(result.reason, RateLimitReason::Whitelisted));
    }
}